src/measure_theory/pi_system.lean

```diff
@@ -0,0 +1,188 @@
+
+* `generate_pi_system g` gives the minimal pi system containing `g`.
+  This can be considered a Galois insertion into both measurable spaces and sets.
+
+* `generate_from_generate_pi_system_eq` proves that if you generate a pi_system
+  then a measurable space versus generating a measurable space gives the same result. This
+  is useful because there are connections between a independent sets that are pi systems
+  the generated independent spaces.
+
+* `mem_generate_pi_system_Union_elim` and `mem_generate_pi_system_Union_elim'` are theorems
+  that show that any element of the supremum of the union of a set of pi systems can be
+  represented as the intersection of a finite number of elements from these sets.
+
+## Implementation details
+
+* is_pi_system is a predicate, not a type. Thus, we don't explicitly define the galois
+  insertion, nor do we define a complete lattice. In theory, we could define a complete
+  lattice and galois insertion on subtype is_pi_system.
+
+-/
+
+open measure_theory measurable_space
+open_locale classical
+
+lemma is_pi_system.singleton {α} (S : set α) : is_pi_system ({S} : set (set α)) :=
+begin
+  intros s t h_s h_t h_ne,
+  rw [set.mem_singleton_iff.1 h_s, set.mem_singleton_iff.1 h_t, set.inter_self,
+      set.mem_singleton_iff],
+end
+
+inductive generate_pi_system {α} (g : set (set α)) : set (set α)
+| base {s : set α} (h_s : s ∈ g) : generate_pi_system s
+| inter {s t : set α} (h_s : generate_pi_system s)  (h_t : generate_pi_system t)
+  (h_nonempty : (s ∩ t).nonempty) : generate_pi_system (s ∩ t)
+
+lemma is_pi_system_generate_pi_system {α} (g : set (set α)) :
+  is_pi_system (generate_pi_system g) :=
+  λ s t h_s h_t h_nonempty, generate_pi_system.inter h_s h_t h_nonempty
+
+lemma generate_pi_system_subset {α} {g t : set (set α)} (h_t : is_pi_system t)
+  (h_sub : g ⊆ t) : (generate_pi_system g : (set (set α))) ⊆ t := begin
+  rw set.subset_def,
+  intros x h,
+  induction h with s h_s s u h_gen_s h_gen_u h_nonempty h_s h_u,
+  apply h_sub h_s,
+  apply h_t _ _ h_s h_u h_nonempty
+end
+
+lemma generate_pi_system_eq {α} {g : set (set α)} (h_pi : is_pi_system g) :
+  generate_pi_system g = g := begin
+  apply le_antisymm,
+  apply generate_pi_system_subset,
+  apply h_pi,
+  apply set.subset.refl,
+  apply generate_pi_system.base,
+end
+
+lemma generate_pi_system_measurable_set {α} [M : measurable_space α] {g : set (set α)}
+  (h_meas_g : ∀ s ∈ g, measurable_set s) (t : set α)
+  (h_in_pi : generate_pi_system g t) : measurable_set t :=
+begin
+  induction h_in_pi with s h_s s u h_gen_s h_gen_u h_nonempty h_s h_u,
+  apply h_meas_g _ h_s,
+  apply measurable_set.inter h_s h_u,
+end
+
+lemma generate_from_measurable_set_of_generate_pi_system {α} {g : set (set α)} (t : set α) :
+  generate_pi_system g t → (measurable_space.generate_from g).measurable_set' t := begin
+  apply generate_pi_system_measurable_set,
+  intros s h_s_in_g,
+  apply measurable_space.measurable_set_generate_from h_s_in_g,
+end
```

Comment:
The preferred style is to put all hypotheses before the ` : `.

Since you opened `measurable_space`, you can write ` generate_from_le` instead of ` measurable_space.generate_from_le`. You should delete those `measurable_space.` in all the proofs.

The suggestion below also uses a term mode proof. Use it or keep you tactic mode proof, as you want.
```suggestion
lemma generate_from_measurable_set_of_generate_pi_system {α} {g : set (set α)} (t : set α)
  (ht : t ∈ generate_pi_system g) :
  (generate_from g).measurable_set' t :=
@generate_pi_system_measurable_set α (generate_from g) g
  (λ s h_s_in_g, measurable_set_generate_from h_s_in_g) t ht
```